src/vxworks/mod.rs

```diff
@@ -12,6 +12,9 @@ impl ::Clone for DIR {
     }
 }
 
+// Throughout we use usize / isize for types that are
+// (unsigned) int in ILP32 and (unsigned) long in ILP64
```

Comment:
`usize == uintptr_t` and `isize == intptr_t` 

`unsigned int` is `c_uint` and `unsigned long` is `c_ulong`.

Reply:
You can use `cfg(target_pointer_width = "...")` to use different types for different pointer widths.

Reply:
I am adding the following code:

cfg_if! {
    if #[cfg(target_pointer_width = "32")] {
        pub type size_t = u32;
        pub type ssize_t = i32;
        pub type ptrdiff_t = i32;
        pub type intptr_t = i32;
        pub type uintptr_t = u32;
    } else if #[cfg(target_pointer_width = "64")] {
        pub type size_t = u64;
        pub type ssize_t = i64;
        pub type ptrdiff_t = i64;
        pub type intptr_t = i64;
        pub type uintptr_t = u64;
    } else {
        // Unknown target_pointer_width                                                                                              
    }
}

But when run the build script (ci/build.sh), I get such error:

error[E0308]: mismatched types
    --> src/vxworks/mod.rs:2156:45
     |
2156 |     if (align & align - 1 != 0) || (align % size_of::<::size_t>() != 0) {
     |                                             ^^^^^^^^^^^^^^^^^^^^^ expected u32, found usize

error[E0277]: cannot mod `u32` by `usize

I think this should be the expected result as size_t was defined as usize and right now it is defined as u32.

Reply:
> `usize == uintptr_t` and `isize == intptr_t`
> 
> `unsigned int` is `c_uint` and `unsigned long` is `c_ulong`.

Understood but maybe we just need to clarify the comment rather than changing the definitions of `size_t `et al.

`size_t = usize`, `ssize_t = isize`, ec, are both layout and "in spirit" compatible with VxWorks. I think it will create a lot of unneeded headaches if we change the definitions of these types.

We have recently got the libc-test working on VxWorks and from what I'm told, this "discrepancy" did not show up as an issue.

Furthermore, the [Unix libc](https://github.com/rust-lang/libc/blob/master/src/unix/mod.rs#L19) provides a similar definition of `size_t` etc. in terms of `usize/isize`, even though e.g. on my Linux box the system header definitions of these types are similar to those in VxWorks.

"Morally" these C types are supposed to act like `usize/isize`. E.g. in Rust `mem::size_of` returns something that  should act like C's `size_t,` right? So it makes sense for them to have the same type (assuming of course that the "real" `size_t` is layout compatible with `usize`, which for our targets it is).

Reply:
Just to confirm, from our libc-test logs with the previous "`size_t = usize`" code:

```
[58] size_t size => Passed
[59] size_t align => Passed
[60] size_t signed => Passed
```

(same result on both 32-bit and 64-bit targets).

Reply:
> align % size_of::<::size_t>()

The problem is that in that line, `align` and `size_of::<...>()` have different types (u32 vs usize), but `%` only works when they have the same size. There are many ways to fix this error, depending on what exactly does the C code do. One way is just adding a cast `align as usize % size_of::<::size_t>()`.

Reply:
> Understood but maybe we just need to clarify the comment rather than changing the definitions of size_t et al.

Yes, my original comment was exclusively about the comment. I was just warning you that `usize` and `isize` do not in general map to `unsigned int`/`int` in C - we use `libc::c_uint` and `libc::c_int` for those throughout the code instead.

Reply:
So if you have a C struct that uses `int`, you should write here `libc::c_int`, instead of using `isize` for that.

Reply:
> I was just warning you that usize and isize do not in general map to unsigned int/int in C - we use libc::c_uint and libc::c_int for those throughout the code instead.

Right, I think we're in agreement, but the comment was poorly written (my fault). In fact it really only applies to types like `size_t` where (I think) it really does make sense to map them to `usize/isize` (as argued above).

We are correctly using `libc::c_int` etc. for fields whose type really is supposed to be a C `int`.

I suggest moving the comment above the block of code that defines size_t et al, and changing it to:

```
// size_t and similar size types are defined in VxWorks as (unsigned) int
// in ILP32 and (unsigned) long in ILP64, but it is both layout compatible
// and "in the spirit" of the VxWorks definitions and C semantics to define them in
// terms of usize/isize here.
```

Reply:
FWIW my recommendation here is try to stick to using the same types as C. Sometimes, using an u32 when C uses an int will work, but is much clearer for everybody to just translate the C sources to Rust "as is".

So if C defines `size_t` as unsigned int, just write:

```rust
type size_t = c_uint;
```

if `c_uint` is the same size of a pointer, you can then also write:

```rust
type c_uint = uintptr_t;
```

Reply:
That way, it all just works out nicely, and you don't need a documentation comment.

Reply:
@gnzlbg 
@n-salim

Please help to check my latest update and let me if that's what we want.